src/editor/dag.rs

```diff
@@ -82,10 +82,15 @@ pub enum EditErr {
     NoChangesToRedo,
     /// The user typed a char that doesn't correspond to any node
     CharNotANode(char),
+    /// Trying to insert a node that cannot be root node
+    CannotBeRoot {
+        /// The [`str`] representing the disallowed node type
+        node_type: &'static str,
+    },
```

Comment:
I think here it would be good to use a `NodeClass`, because we can make use of the extra information.  It would require a type parameter, though:
```rust
pub enum EditErr<C: NodeClass> {
    CannotBeRoot {
        /// The class of the dis-allowed node
        class: C
    },
}
```

Reply:
I'm not sure how to implement this. 
```rust
    fn insert_next_to_cursor(&mut self, c: char, side: Side) -> EditResult {
        let node_type = Node::Class::from_char(c).ok_or(EditErr::CharNotANode(c))?;
        self.perform_edit(
            |this: &mut Self,
             parent_and_index: Option<(&'arena Node, usize)>,
             _cursor: &'arena Node| {
                let (parent, cursor_index) = parent_and_index.ok_or(EditErr::AddSiblingToRoot)?;

                // Short circuit if not an insertable char
                if !parent.is_valid_child(cursor_index, c) {
                if !parent.is_valid_child(cursor_index, node_type) {
                    return Err(EditErr::CannotBeChild {
                        class: note_type, // class has AstClass, but node_type has type Ast::Class
                        node_type: node_type.name(),
                        parent_name: parent.display_name(),
                    });
                }
```
Complier doesn't  think `Ast::Class` has trait bound `AstClass`

Reply:
> Complier doesn't think `Ast::Class` has trait bound `AstClass`

That's odd.  It should - because `Node::Class` is required to implement `AstClass`.  I also managed to get this work on my local machine.  This is a rough sketch of what I had:

### src/editor/dag.rs
```rust
pub enum EditSuccess<C: AstClass> {
    Undo,
    Redo,
    Move(Direction),
    Replace(C),
    InsertChild(C),
    InsertNextToCursor { side: Side, class: C },
    Delete { name: String },
}

pub enum EditErr<C: AstClass> {
    ...
    /// Trying to insert a node that cannot be root node
    CannotBeRoot(C),
    ...
}

impl Dag<'arena, Node: Ast<'arena>> {
    ...

    fn insert_child(&mut self, c: char) -> EditResult<Node::Class> {
        let class = Node::Class::from_char(c).ok_or(EditErr::CharNotANode(c))?;
        self.perform_edit(
            |this: &mut Self,
                ...

                Ok((
                    cloned_cursor,
                    EditLocation::Cursor,
                    // This line was changed - we just insert `class` directly instead of getting its name and char
                    EditSuccess::InsertChild(class),
                 ))
            },
        )
    }
    ...
}
```

> Is trait `AstClass` required to have lifetime `'arena`? ... And we need have lifetime `'arena` in `EditErr` definition. Complier will complain: unused parameter `'arena`.

We don't need the `<'arena>` lifetime annotation in `AstClass`, because lifetime annotations specify the lifetime of any data borrowed by the datatype - but `AstClass` should be a standalone enum, so we don't want it to borrow data.

Reply:
I finally made it work. LOL

Reply:
Nice! Rust's compiler for all its greatness can be really silly sometimes :laughing:.